core/src/main/java/hudson/model/Run.java

```diff
@@ -1930,31 +1921,55 @@ public synchronized void save() throws IOException {
      * @throws IOException If there is a problem reading the log file.
      */
     public @Nonnull List<String> getLog(int maxLines) throws IOException {
-        int lineCount = 0;
-        List<String> logLines = new LinkedList<String>();
-        if (maxLines == 0) {
-            return logLines;
-        }
-        try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(getLogFile()), getCharset()))) {
-            for (String line = reader.readLine(); line != null; line = reader.readLine()) {
-                logLines.add(line);
-                ++lineCount;
-                // If we have too many lines, remove the oldest line.  This way we
-                // never have to hold the full contents of a huge log file in memory.
-                // Adding to and removing from the ends of a linked list are cheap
-                // operations.
-                if (lineCount > maxLines)
-                    logLines.remove(0);
+        final List<String> lastLines = new ArrayList<>(maxLines);
+
+        if(maxLines == 0) {
+            return lastLines;
+        }
+
+        int lines = 0;
+        final List<Byte> bytes = new ArrayList<>();
+        try (RandomAccessFile fileHandler = new RandomAccessFile(getLogFile(), "r")) {
+            long fileLength = fileHandler.length() - 1;
+
+            for (long filePointer = fileLength; filePointer != -1 && maxLines != lines; filePointer--) {
+                fileHandler.seek(filePointer);
+                byte readByte = fileHandler.readByte();
+
+
+                if (readByte == 0x0A) {
+                    if (filePointer < fileLength) {
+                        lines = lines + 1;
+                        lastLines.add(convertBytesToString(bytes));
+                        bytes.clear();
+                    }
+                }
+                else if (readByte != 0xD) {
+                    bytes.add(readByte);
+                }
             }
         }
 
+        if (lines != maxLines) {
+            lastLines.add(convertBytesToString(bytes));
+        }
+
+        Collections.reverse(lastLines);
+
         // If the log has been truncated, include that information.
         // Use set (replaces the first element) rather than add so that
         // the list doesn't grow beyond the specified maximum number of lines.
-        if (lineCount > maxLines)
-            logLines.set(0, "[...truncated " + (lineCount - (maxLines - 1)) + " lines...]");
+        if (lines == maxLines) {
+            lastLines.set(0, "[...truncated lines...]");
+        }
+
+        return ConsoleNote.removeNotes(lastLines);
+    }
 
-        return ConsoleNote.removeNotes(logLines);
+    private String convertBytesToString(List<Byte> bytes) {
+        Collections.reverse(bytes);
```

Comment:
IIRC there is a method, which creates new collection from ArrayList or LinkedList without updating the original one. Should be better from the performance PoV. Or you can just use Reverese iterator since you have an Array list